File: document/src/vespa/document/annotation/annotation.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "annotation.h"
#include "spannode.h"
#include <vespa/document/fieldvalue/fieldvalue.h>
#include <vespa/document/util/bytebuffer.h>
#include <vespa/vespalib/stllike/asciistream.h>

namespace document {

std::ostream & operator << (std::ostream & os, const Annotation &annotation) {
    return os << annotation.toString();
}

Annotation::~Annotation() = default;

vespalib::string
Annotation::toString() const {
    vespalib::asciistream os;
    os << "Annotation(" << *_type;
    if (_value.get()) {
        os << "\n";
        os << _value->toString();
    }
    if (_node) {
        os << "\n";
        os << _node->toString();
    }
    os << ")";
    return os.str();
}

bool
Annotation::operator==(const Annotation &a2) const {
    return (getType() == a2.getType() &&
            !(!!getFieldValue() ^ !!a2.getFieldValue()) &&
            (!getFieldValue() || (*getFieldValue() == *a2.getFieldValue()))
           );
}

}  // namespace document

File: client/src/main/java/ai/vespa/client/dsl/Annotation.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package ai.vespa.client.dsl;

import java.util.HashMap;
import java.util.Map;

public class Annotation {

    private final Map<String, Object> annotations;

    Annotation() {
        this(new HashMap<>());
    }

    Annotation(Map<String, Object> annotations) {
        this.annotations = annotations;
    }

    public Annotation append(Annotation a) {
        this.annotations.putAll(a.annotations);
        return this;
    }

    public boolean contains(String key) {
        return annotations.containsKey(key);
    }

    public Object get(String key) {
        return annotations.get(key);
    }

    @Override
    public String toString() {
        return annotations == null || annotations.isEmpty() ? "" : Q.toJson(annotations);
    }

}
